src/rvary_fmt_plug.c

```diff
@@ -0,0 +1,325 @@
+	  19,  16,  22,  21, 127, 124, 115, 112, 118, 117,  94,  93,  82,  81,  87,  84
+};
+
+/* in-place ascii2ebcdic conversion */
+static void ascii2ebcdic(unsigned char *str)
+{
+	int i;
+	int n = strlen((const char*)str);
+	for (i = 0; i < n; ++i)
+		str[i] = a2e[str[i]];
+}
+
+/* Pad input string with EBCDIC spaces (0x40) up to eight bytes */
+static void ebcdic_padding(unsigned char *str)
+{
+	int i;
+	for (i = strlen((const char*)str); i < 9; ++i)
+		str[i] = 0x40;
+	str[9] = 0; /* terminate string */
+}
+
+#ifdef RVARY_DEBUG
+static void print_hex(unsigned char *str, int len)
+{
+	int i;
+	for (i = 0; i < len; ++i)
+		printf("%02x", str[i]);
+	printf("\n");
+}
+#endif
+
+static struct fmt_tests rvary_tests[] = {
+	{"$rvary$**5AA70358A9C369E0", "QWERTY1"},
+	{"$rvary$**EB59EE07FC74AE77", "PASSWORD"},
+	{"$rvary$**062314297C496E0E", "AAAAAAAA"},
+	{"$rvary$**0FF48804F759193F", "TESTTEST"},
+	{NULL}
+};
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static uint32_t (*crypt_out)[BINARY_SIZE / sizeof(uint32_t)];
+static DES_key_schedule (*schedules);
+static int dirty;
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+
+	saved_key = mem_calloc(self->params.max_keys_per_crypt,
+	                       sizeof(*saved_key));
+	crypt_out = mem_calloc(self->params.max_keys_per_crypt,
+	                       sizeof(*crypt_out));
+	schedules = mem_calloc(self->params.max_keys_per_crypt,
+	                       sizeof(*schedules));
+}
+
+static void done(void)
+{
+	MEM_FREE(schedules);
+	MEM_FREE(crypt_out);
+	MEM_FREE(saved_key);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *ctcopy;
+	char *keeptr;
+	int extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LEN))
+		return 0;
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LEN;
+	ctcopy += 1; /* Skip empty salt */
+	if (hexlenu(ctcopy, &extra) != CIPHERTEXT_LENGTH || extra)
+		goto err;
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_binary(char *ciphertext)
+{
+	static union {
+		unsigned char c[BINARY_SIZE+1];
+		ARCH_WORD dummy;
+	} buf;
+	unsigned char *out = buf.c;
+	char *p;
+	int i;
+
+	p = strrchr(ciphertext, '*') + 1;
+		for (i = 0; i < BINARY_SIZE; i++) {
+		out[i] =
+		    (atoi16[ARCH_INDEX(*p)] << 4) |
+		    atoi16[ARCH_INDEX(p[1])];
+		p += 2;
+	}
+
+	return out;
+}
+
+#define COMMON_GET_HASH_VAR crypt_out
+#include "common-get-hash.h"
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index++) {
+		if (dirty) {
+			DES_cblock des_key;
+			int i;
+
+			/* process key */
+			for (i = 0; saved_key[index][i]; i++)
+				des_key[i] = a2e_precomputed[ARCH_INDEX(saved_key[index][i])];
+
+			/* replace missing characters in userid by (EBCDIC space (0x40) XOR 0x55) << 1 */
+			while(i < 9)
+				des_key[i++] = 0x2a;
```

Comment:
This is where ASan complains. The `while` loop allows for `i` up to 8 inclusive, but `des_key` is a `DES_cblock` and I guess is exactly 8 bytes. So we write one byte beyond the buffer end here.

Reply:
And indeed `racf_fmt_plug.c` has the check as `< 8`. Where did the 9 come from?

Reply:
I see. When RACF encrypts the user password with the "traditional" DES implementation, it's using the username (up to 7 chars) as salt.
RVARY instead encrypts the password with the password itself as salt. Easy fix then.

Reply:
Fixed in aa66ef1